oak_restricted_kernel/src/syscall/mod.rs

```diff
@@ -0,0 +1,155 @@
+//
+// Copyright 2022 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+mod channel;
+
+use alloc::boxed::Box;
+use core::{arch::asm, ffi::c_void};
+use oak_channel::Channel;
+use oak_restricted_kernel_runtime::{syscall::Errno, Syscall};
+use x86_64::{
+    registers::{
+        control::Efer,
+        model_specific::{EferFlags, LStar},
+    },
+    VirtAddr,
+};
+
+use self::channel::{syscall_fsync, syscall_read, syscall_write};
+
+pub fn enable_syscalls(channel: Box<dyn Channel>) {
+    channel::init(channel);
+
+    LStar::write(VirtAddr::new(syscall_entrypoint as *const fn() as u64));
+    unsafe {
+        Efer::update(|flags| flags.set(EferFlags::SYSTEM_CALL_EXTENSIONS, true));
+    }
+}
+
+// SysV ABI: arguments are in RDI, RSI, RDX, RCX; return value in RAX (which matches SYSRET!)
+extern "sysv64" fn syscall_handler(syscall: usize, arg1: usize, arg2: usize, arg3: usize) -> isize {
```

Comment:
What does this do?
I know that extern C prevents name mangling and exposes the function for linkage, but that does the `extern "sysv64"` do?

Reply:
It's more than about name manging -- this specifies the ABI of the function (how arguments are passed to the function and how the return value is passed back, for example).

See https://doc.rust-lang.org/reference/items/external-blocks.html#abi for more details.

Under Linux, `extern "C"` is the same as `extern "sysv64"` (on Windows, `extern "C"` maps to `extern "win64"`).

For us, we're building for `x86_64-unknown-none` target, which also uses sysv64. This just explicitly states that fact.